jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EventsHandler.java

```diff
@@ -0,0 +1,419 @@
+ * are always passed. If somewhere in the parent chain is the {@code GzipHandler} then
+ * unzipped chunks are passed.</p>
+ */
+public abstract class EventsHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(EventsHandler.class);
+    
+    public EventsHandler()
+    {
+    }
+
+    public EventsHandler(Handler handler)
+    {
+        super(handler);
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        // Before handling
+        ReadOnlyRequest roRequest = new ReadOnlyRequest(request);
+        fireOnBeforeHandling(roRequest);
+        try
+        {
+            EventsRequest wrappedRequest = new EventsRequest(request, roRequest);
+            EventsResponse wrappedResponse = new EventsResponse(roRequest, response);
+            request.addHttpStreamWrapper(stream -> new HttpStream.Wrapper(stream)
+            {
+                private void fireNotifications(Throwable failure)
+                {
+                    if (!wrappedResponse.notifiedOnResponseBegin)
+                        fireOnResponseBegin(roRequest, wrappedResponse.getHeaders().asImmutable(), wrappedResponse.getStatus());
+                    if (wrappedResponse.suppliedTrailers != null)
+                        fireOnResponseTrailersComplete(roRequest, wrappedResponse.suppliedTrailers);
+                    fireOnComplete(roRequest, failure);
+                }
+
+                @Override
+                public void succeeded()
+                {
+                    fireNotifications(null);
+                    super.succeeded();
+                }
+
+                @Override
+                public void failed(Throwable x)
+                {
+                    fireNotifications(x);
+                    super.failed(x);
+                }
+            });
+
+            boolean handled = super.handle(wrappedRequest, wrappedResponse, callback);
+
+            fireOnAfterHandling(roRequest, handled, null);
+            return handled;
+        }
+        catch (Throwable x)
+        {
+            fireOnAfterHandling(roRequest, false, x);
+            throw x;
+        }
+    }
+
+    private void fireOnBeforeHandling(Request request)
+    {
+        try
+        {
+            onBeforeHandling(request);
+        }
+        catch (Throwable x)
+        {
+            LOG.info("Error firing onBeforeHandling", x);
+        }
+    }
+
+    private void fireOnRequestRead(Request wrapped, Content.Chunk chunk)
+    {
+        try
+        {
+            Content.Chunk c;
+            if (chunk == null)
+                c = null;
+            else if (chunk.hasRemaining())
+                c = Content.Chunk.asChunk(chunk.getByteBuffer().asReadOnlyBuffer(), chunk.isLast(), chunk);
+            else
+                c = chunk.isLast() ? Content.Chunk.EOF : Content.Chunk.EMPTY;
```

Comment:
Is this necessary? why not just pass the chunk:
```suggestion
                c = chunk;
```

Reply:
I've simplified this logic and moved it to `Chunk.asReadOnly()`.